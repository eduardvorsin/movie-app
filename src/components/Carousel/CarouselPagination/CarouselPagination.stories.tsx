import type { Meta, StoryObj } from '@storybook/react';
import CarouselPagination, { Props } from './CarouselPagination';
import { useTranslation } from 'react-i18next';
// eslint-disable-next-line import/no-extraneous-dependencies
import { action } from '@storybook/addon-actions';

const meta: Meta<typeof CarouselPagination> = {
	title: 'UI/Carousel/CarouselPagination',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		testId: {
			description: 'Id for testing the component',
		},
		activeIndex: {
			description: 'Index of the active pagination element',
		},
		totalCount: {
			description: 'The total number of pagination elements',
		},
		paginationType: {
			description: 'Type of pagination',
		},
		onDotClick: {
			control: false,
			description: 'Allows you to send a callback that will be called when clicked',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
	},
	component: CarouselPagination,
	parameters: {
		docs: {
			description: {
				component: 'The CarouselPagination component is a control used to display and select specific pages in the carousel component. This component allows users to quickly navigate between different sets of content inside the carousel.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof CarouselPagination>;

const CarouselPaginationWithHooks = (props: Omit<Props, 'dictionary' | 'onDotClick'>) => {
	const { t } = useTranslation('common');
	const dictionary = { label: t('carouselPagination.label') };
	const dotClickHandler = () => action('Clicked')();

	if (props.paginationType === 'dots') {
		return (
			<CarouselPagination
				{...props}
				paginationType={'dots'}
				dictionary={dictionary}
				onDotClick={dotClickHandler}
			/>
		);
	}

	return (
		<CarouselPagination
			{...props}
			paginationType={props.paginationType === 'progress' ? 'progress' : 'fraction'}
		/>
	);
}

export const Default: Story = {
	args: {
		activeIndex: 3,
		totalCount: 10,
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (<CarouselPaginationWithHooks {...args} />),
};

export const PaginationType: Story = {
	args: {
		activeIndex: 2,
		totalCount: 5,
	},
	parameters: {
		docs: {
			description: {
				story: 'The type of pagination. There are 3 types of pagination available: "dots", "fraction", "progress".',
			},
		},
	},
	render: (args) => (
		<div className='flex flex-col gap-5 max-w-[200px]'>
			<CarouselPaginationWithHooks {...args} paginationType='dots' />
			<CarouselPaginationWithHooks {...args} paginationType='progress' />
			<CarouselPaginationWithHooks {...args} paginationType='fraction' />
		</div>
	),
};

export const EventCallbacks: Story = {
	args: {
		paginationType: 'dots',
		activeIndex: 4,
		totalCount: 7,
	},
	parameters: {
		docs: {
			description: {
				story: 'The callback that is called when clicking on the pagination element',
			},
		},
	},
	render: (args) => (<CarouselPaginationWithHooks {...args} />),
};