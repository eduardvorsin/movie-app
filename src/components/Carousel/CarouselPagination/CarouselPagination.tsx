'use client';
import { GeneralProps } from '@/types/shared';
import { MouseEventHandler, useRef } from 'react';

type Props = {
	activeIndex: number,
	totalCount: number,
	onDotClick: (index: number) => void,
	paginationType: 'dots' | 'progress' | 'fraction',
} & GeneralProps;

export default function CarouselPagination({
	className,
	testId,
	onDotClick,
	paginationType,
	activeIndex,
	totalCount,
	...props
}: Props) {

	const dotsContainerRef = useRef<HTMLDivElement>(null);

	const clickHandler: MouseEventHandler<HTMLButtonElement | HTMLDivElement> = (e) => {
		const target = e.target as HTMLButtonElement | HTMLDivElement;
		const dotsContainerNode = dotsContainerRef.current;
		if (!onDotClick || !dotsContainerNode || target.tagName !== 'BUTTON') {
			return;
		}

		const index = Array.from(dotsContainerNode.children).findIndex((dotElement) => dotElement === target);
		onDotClick(index);
	};

	const classes = [
		'w-full flex justify-center items-center text-center text-200 text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150',
		paginationType === 'progress' ? 'h-1 bg-neutral-300 dark:bg-dark-neutral-300' : '',
		className,
	].join(' ');

	let Pagination: JSX.Element;

	if (paginationType === 'fraction') {
		Pagination = (
			<>
				<span>{activeIndex + 1}</span>
				<span className='mx-1'>/</span>
				<span>{totalCount < 1 ? 1 : totalCount}</span>
			</>
		);
	} else if (paginationType === 'dots') {
		const dots = new Array(totalCount).fill(null);
		const activeDotClasses = 'border-blue-600 hover:border-blue-700 active:border-blue-800 dark:border-blue-300 dark:hover:border-blue-400 dark:active:border-blue-500 bg-transparent';
		const nonActiveDotClasses = 'bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500';

		Pagination = (
			//event bubbling 
			// eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions
			<div
				className='flex py-1 gap-2 overflow-x-auto'
				ref={dotsContainerRef}
				onClick={clickHandler}
				role='group'
				aria-label='carousel pagination'
			>
				{dots.map((_, index) => (
					<button
						key={index}
						className={`w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 ${index === activeIndex ? activeDotClasses : nonActiveDotClasses} `}
						aria-disabled={index === activeIndex}
					/>
				))}
			</div>
		);
	} else {
		Pagination = (
			<span
				className='bg-blue-700 dark:bg-blue-300 w-full h-full scale-y-100 origin-top-left transition-[transform,background-color] duration-150'
				style={{
					transform: `scaleX(${(activeIndex + 1) / totalCount})`
				}}
			/>
		);
	}

	return (
		<div
			className={classes}
			data-testid={testId}
			{...props}
		>
			{Pagination}
		</div>
	);
}
