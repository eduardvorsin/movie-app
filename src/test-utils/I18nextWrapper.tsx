import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;