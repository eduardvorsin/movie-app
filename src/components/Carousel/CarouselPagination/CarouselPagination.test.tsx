import { render, screen } from "@testing-library/react";
import CarouselPagination from "./CarouselPagination";
import userEvent from "@testing-library/user-event";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

const dictionary = { label: 'Carousel pagination' }

describe('CarouselPagination tests', () => {
	it('is rendered correctly', () => {
		render(
			<CarouselPagination
				paginationType='progress'
				totalCount={8}
				activeIndex={2}
				testId='test-carousel-pagination'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel-pagination')).toBeInTheDocument();
	});

	it('when clicking on the navigation element, the mock function should be called', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<CarouselPagination
				paginationType='dots'
				totalCount={8}
				activeIndex={2}
				onDotClick={mockFn}
				dictionary={dictionary}
				testId='test-carousel-pagination'
			/>
		);

		await user.click(screen.getAllByRole<HTMLButtonElement>('button')[0]);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel-pagination')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		render(
			<CarouselPagination
				paginationType='progress'
				totalCount={8}
				activeIndex={2}
				testId='test-carousel-pagination'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel-pagination')).toMatchSnapshot();
	});

	it('is a snapshot with paginationType equal to "progress"', () => {
		render(
			<CarouselPagination
				paginationType='progress'
				totalCount={8}
				activeIndex={2}
				testId='test-carousel-pagination'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel-pagination')).toBeInTheDocument();
	});

	it('is a snapshot with paginationType equal to "dots"', () => {
		const mockFn = jest.fn();
		render(
			<CarouselPagination
				paginationType='dots'
				totalCount={8}
				activeIndex={2}
				onDotClick={mockFn}
				dictionary={dictionary}
				testId='test-carousel-pagination'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel-pagination')).toBeInTheDocument();
	});

	it('is a snapshot with paginationType equal to "fraction"', () => {
		render(
			<CarouselPagination
				paginationType='fraction'
				totalCount={8}
				activeIndex={2}
				testId='test-carousel-pagination'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-carousel-pagination')).toBeInTheDocument();
	});
});

describe('CarouselPagination integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('carouselPagination.label'));

		const mockFn = jest.fn();
		const dictionary = { label: i18next.t('carouselPagination.label') }
		render(
			<CarouselPagination
				paginationType='dots'
				totalCount={8}
				activeIndex={2}
				onDotClick={mockFn}
				dictionary={dictionary}
				testId='test-carousel-pagination'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLDivElement>('group', { name: translation })).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('carouselPagination.label'));

		const mockFn = jest.fn();
		const dictionary = { label: i18next.t('carouselPagination.label') }
		render(
			<CarouselPagination
				paginationType='dots'
				totalCount={8}
				activeIndex={2}
				onDotClick={mockFn}
				dictionary={dictionary}
				testId='test-carousel-pagination'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLDivElement>('group', { name: translation })).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('carouselPagination.label'));

		const mockFn = jest.fn();
		const dictionary = { label: i18next.t('carouselPagination.label') }
		render(
			<CarouselPagination
				paginationType='dots'
				totalCount={8}
				activeIndex={2}
				onDotClick={mockFn}
				dictionary={dictionary}
				testId='test-carousel-pagination'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLDivElement>('group', { name: translation })).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('carouselPagination.label'));

		const mockFn = jest.fn();
		const dictionary = { label: i18next.t('carouselPagination.label') }
		render(
			<CarouselPagination
				paginationType='dots'
				totalCount={8}
				activeIndex={2}
				onDotClick={mockFn}
				dictionary={dictionary}
				testId='test-carousel-pagination'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLDivElement>('group', { name: translation })).toMatchSnapshot();
	});
});