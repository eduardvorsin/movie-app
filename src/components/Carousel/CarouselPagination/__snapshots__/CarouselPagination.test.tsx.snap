// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselPagination integration tests is a snapshot with English localization 1`] = `
<div
  aria-label="Carousel pagination"
  class="flex py-1 gap-2 overflow-x-auto"
  role="group"
>
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="true"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 border-blue-600 hover:border-blue-700 active:border-blue-800 dark:border-blue-300 dark:hover:border-blue-400 dark:active:border-blue-500 bg-transparent "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
</div>
`;

exports[`CarouselPagination integration tests is a snapshot with Russian localization 1`] = `
<div
  aria-label="Пагинация слайдера"
  class="flex py-1 gap-2 overflow-x-auto"
  role="group"
>
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="true"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 border-blue-600 hover:border-blue-700 active:border-blue-800 dark:border-blue-300 dark:hover:border-blue-400 dark:active:border-blue-500 bg-transparent "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
  <button
    aria-disabled="false"
    class="w-[0.9375rem] h-[0.9375rem] sm:w-[0.625rem] sm:h-[0.625rem] shrink-0 border-2 rounded-full transition-colors duration-150 bg-neutral-300 hover:bg-neutral-400 active:bg-neutral-500 border-neutral-300 hover:border-neutral-400 active:border-neutral-500 dark:bg-dark-neutral-350 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 dark:border-dark-neutral-350 dark:hover:border-dark-neutral-400 dark:active:border-dark-neutral-500 "
  />
</div>
`;

exports[`CarouselPagination tests is a basic snapshot 1`] = `
<div
  class="w-full flex justify-center items-center text-center text-200 text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150 h-1 bg-neutral-300 dark:bg-dark-neutral-300 "
  data-testid="test-carousel-pagination"
>
  <span
    class="bg-blue-700 dark:bg-blue-300 w-full h-full scale-y-100 origin-top-left transition-[transform,background-color] duration-150"
    style="transform: scaleX(0.375);"
  />
</div>
`;
